runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -110,14 +110,22 @@ local function diagnostic_lsp_to_vim(diagnostics, bufnr, client_id)
   return vim.tbl_map(function(diagnostic)
     local start = diagnostic.range.start
     local _end = diagnostic.range['end']
+    local message ---@type string
+    if type(diagnostic.message) == 'string' then
+      ---@cast diagnostic { message: string }
+      message = diagnostic.message
```

Comment:
LuaLS can't infer based on `type(diagnostic.message) == 'string'` ?

Reply:
Sadly it can't :(

Reply:
`.message` needs to be put into a intermediate variable for it to correctly narrow the type.